internal/impl/kafka/enterprise/schema_registry_input.go

```diff
@@ -0,0 +1,265 @@
+			service.NewStringField(sriFieldSubjectFilter).Description("Include only subjects which match the regular expression filter. All subjects are selected when not set.").Default("").Advanced(),
+			service.NewTLSToggledField(sriFieldTLS),
+			service.NewAutoRetryNacksToggleField(),
+		)
+}
+
+func init() {
+	err := service.RegisterInput("schema_registry", inputSpec(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.Input, error) {
+			i, err := inputFromParsed(conf, mgr.Logger())
+			if err != nil {
+				return nil, err
+			}
+			return service.AutoRetryNacksToggled(conf, i)
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+type input struct {
+	url           *url.URL
+	subjectFilter *regexp.Regexp
+
+	client    http.Client
+	connected atomic.Bool
+	subjects  []string
+	subject   string
+	versions  []int
+	log       *service.Logger
+}
+
+func inputFromParsed(pConf *service.ParsedConfig, log *service.Logger) (i *input, err error) {
+	i = &input{
+		log: log,
+	}
+
+	var u string
+	if u, err = pConf.FieldString(sriFieldURL); err != nil {
+		return
+	}
+	if i.url, err = url.Parse(u); err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %s", err)
+	}
+
+	var includeDeleted bool
+	if includeDeleted, err = pConf.FieldBool(sriFieldIncludeDeleted); err != nil {
+		return
+	}
+	if includeDeleted {
+		q := i.url.Query()
+		q.Add("deleted", "true")
+		i.url.RawQuery = q.Encode()
+	}
+
+	var filter string
+	if filter, err = pConf.FieldString(sriFieldSubjectFilter); err != nil {
+		return
+	}
+	if i.subjectFilter, err = regexp.Compile(filter); err != nil {
+		return nil, fmt.Errorf("failed to compile subject filter %q: %s", filter, err)
+	}
+
+	var tlsConf *tls.Config
+	var tlsEnabled bool
+	if tlsConf, tlsEnabled, err = pConf.FieldTLSToggled(sriFieldTLS); err != nil {
+		return
+	}
+
+	i.client = http.Client{}
+	if tlsEnabled && tlsConf != nil {
+		if c, ok := http.DefaultTransport.(*http.Transport); ok {
+			cloned := c.Clone()
+			cloned.TLSClientConfig = tlsConf
+			i.client.Transport = cloned
+		} else {
+			i.client.Transport = &http.Transport{
+				TLSClientConfig: tlsConf,
+			}
+		}
+	}
+
+	return
+}
+
+//------------------------------------------------------------------------------
+
+func (i *input) Connect(ctx context.Context) error {
+	if i.connected.Load() {
+		return nil
+	}
+
+	data, err := doSchemaRegistryRequest(ctx, i.client, i.url.JoinPath("subjects").String())
+	if err != nil {
+		return fmt.Errorf("failed to fetch subjects: %s", err)
+	}
+
+	var subjects []string
+	if err := json.Unmarshal(data, &subjects); err != nil {
+		return fmt.Errorf("failed to unmarshal HTTP response: %s", err)
+	}
+
+	i.subjects = make([]string, 0, len(subjects))
+	for _, s := range subjects {
+		if i.subjectFilter.MatchString(s) {
+			i.subjects = append(i.subjects, s)
+		}
+	}
+
+	i.connected.Store(true)
+
+	return nil
+}
+
+func (i *input) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
```

Comment:
Is this only going to have a single goroutine calling it?

Reply:
It's a good discussion topic for the next Blobtalk. We really should document somehow what the state machine is. Like, can `Connect()` be called in parallel by multiple goroutines and why? For `Read()`, I guess one can define a `schema_registry` input component and reference it in multiple streams or use a `broker` input with `copies > 1`, so all bets are off...